<!-- \VignetteEngine{knitr::knitr} -->

# twoStageGSA: Two Stage Gene Set Analysis

This packages provides R functions to perform ["A Flexible Two Stage
Procedure for Identifying Gene Sets that are Differentially Expressed"
(Heller, et al. 2009)][hel09]


## A quick example

First we load `twoStageGSA`. We will also make use of pathway database
`reactome.db` and annotation database `hgu95av2.db`.

```{r}
library(twoStageGSA)
library(reactome.db)
library(hgu95av2.db)
library(KEGG.db)
library(GO.db)
library(pander)
```

```{r}
m <- 6
G <- 1000
data <- matrix(rnorm(m*G),ncol= m) +
  cbind(matrix(0,nc=m/2,nr=G),matrix(rep(rnorm(G,sd=2/3),3),nc=3))
labels <- rep(c(0,1),each=m/2)

```

To set up the example we simulate expressions of `r G` genes from `r
m` samples. We assume two distinct groups and add some effects to one
the samples in one of them. 

To make annotations work we assign random probe ids from Affymetrix'
hgu95av2 microarray as rownames of our dataset and randomly draw
100 genesets of different lengths from those. 

```{r}
genes <- keys(hgu95av2SYMBOL,keytype="ENTREZID")
rownames(data) <- sample(genes,G)
geneSets <- sapply(sample(10:100,100,rep=T),function(n) sample(rownames(data),n))
pathways <- keys(reactome.db,keytype="PATHID")
names(geneSets) <- sample(pathways,length(geneSets))
```

``geneSets`` is just a list of character vectors containing the probe
IDs making up the gene sets.



```{r}
geneSets[[1]][1:10]
```

``data`` is a matrix where each line corresponds to a probe and each
column to a sample. Here, the row names correspond to the probe IDs. 

```{r}
head(data)

```

Next we compute mean differences between groups. This will be used to
determine the direction of the effect.

```{r}
diff <- rowMeans(data[,labels==1])-rowMeans(data[,labels==0])
```

We perform the screening step using the `globalTest` to test each
geneset for enrichment. 

```{r}
out <- screening(data,labels,geneSets,B=0,settest='globalTest')
```
We then use the function `t.test` for single gene tests and adjust
form multiplicity using the Bonferroni Holm procedure.

```{r}
fc <- focus(data,labels,geneSets,out$sigSets,test='t.test',adj='holm')
```

To produce nice reports we have to pass appropriate annotation
databases to the reporting functions. 

```{r}

screen <- reportScreening(out,pw.annotation = reactome.db,
                          pw.kv = c('PATHID','PATHNAME'))

focus <- reportFocus(out,fc,diff,q=.5,
                     pw.annotation=reactome.db,
                     gene.annotation = hgu95av2.db,
                     pw.kv = c('PATHID','PATHNAME'),
                     gene.kv = c('PROBEID','SYMBOL'))
```

```{r screening_result,results='markup'}
pander(screen)

```

```{r focus_result,results='markup'}
pander(focus)

```
## Different global test procedures for the screening step

``twoStageGSA' provides several global test procedures to test the
intersection null hypotheses that there is no difference in gene
expression between genes in a gene set.

### Global test [goe04]

### Hotelling T-test 

### Nettleton [net08]


## Writing custom screening test procedures 

## Writing custom focus test procedures

### Re-using results from previous analyses

In case that a single gene differential analysis has already been
performed, it may be reasonable to reuse the outcome of this
analysis. For the focus step it is natural to use the same test
procedure that one would use in a single gene analysis. If that has
already been performed on ``twoStageGSA`` permits to use the
un-adjusted p-values.

```{r}
myps <- apply(data,1,function(x) t.test(x~labels)$p.value)

screening_myminps <- function(set,perm,d,l){
    minp <- min(myps[set])
}
    
out <- screening(data,labels,geneSets,B=0,settest='myminps')

fc <- focus(data,labels,geneSets,out$sigSets,test=myps,adj='holm',reuse.p=T)

focus <- reportFocus(out,fc,diff,q=.5,
                     pw.annotation=reactome.db,
                     gene.annotation = hgu95av2.db,
                     pw.kv = c('PATHID','PATHNAME'),
                     gene.kv = c('PROBEID','SYMBOL'))

```

```{r focus_results_reuse}
pander(focus)

```
### Inverse normal combination test

### Simes test

### Bonferroni-Holm test




[goe04]:
http://bioinformatics.oxfordjournals.org/content/20/1/93.short "Goeman, J. J., Van De Geer, S. A., De Kort, F., & Van Houwelingen, H. C. (2004). A global test for groups of genes: testing association with a clinical outcome. Bioinformatics, 20(1), 93-99."
[net08]:
http://bioinformatics.oxfordjournals.org/content/24/2/192.short "Nettleton, D., Recknor, J., & Reecy, J. M. (2008). Identification of differentially expressed gene categories in microarray studies using nonparametric multivariate analysis. Bioinformatics, 24(2), 192-201."
[hel09]:
http://bioinformatics.oxfordjournals.org/content/25/8/1019.short "Heller, R., Manduchi, E., Grant, G. R., & Ewens, W. J. (2009). A flexible two-stage procedure for identifying gene sets that are differentially expressed. Bioinformatics, 25(8), 1019-1025."
